api-proposal.md

```diff
@@ -0,0 +1,138 @@
+# Proposal for ecosystem website api.
+
+## API's
+
+---
+
+### Package List
+
+We could have some specifc string matching for categories and tags in the query if thats what the user is searching for to maybe speed things up. `cat:<category>` and `tag:<string>`.
+
+- GET `/api/v1/packages?query=:string&page=:number`
+  ```
+    items: [<package>]
+  ```
+
+---
+
+### Package
+
+- **Schema**
+
+  ```
+  <package> {
+    id: number
+    name: string
+    description: string
+    readme: string // maybe from github or something?
+    image: blob || string
+    website: string
+    repository: string
+    license: string
+    commentsCount: number
+    upvotes: number
+    downvotes: number
+    tags: string[]
+    added: datetime
+    updated: datetime
+  }
+
+  ```
+
+- GET `/api/v1/package/:package`
+  ```
+  data: {
+    package: <package>
+    comments: [<comment>] // first 50(?) comments
+    releases: [<release>] // first 10(?) releases
+  }
+  ```
+- POST `/api/v1/package`
+  ```
+  data: <package>
+  ```
+- PATCH `/api/v1/package/:package`
+  ```
+  data: <package>
+  ```
+
+#### Package Comments
+
+- **Schema**
+
+  ```
+  <comment> {
+    id: number
+    username: string // github username
+    text: string
+    added: datetime
+    updated: datetime
+  }
+  ```
+
+- GET `/api/v1/package/:package/comments?page=:number`
+  ```
+  data: [<comment>] // (x50)
+  ```
+- POST `/api/v1/package/:package/comments`
+  ```
+  data: <comment>
+  ```
+- PATCH `/api/v1/package/:package/comments/:id`
+  ```
+  data: <comment>
+  ```
+- DELETE `/api/v1/package/:package/comments/:id`
+
+#### Package Releases
+
+- **Schema**
+
+  ```
+  <release> {
+    id: number
+    number: string //(semver probably)
+    description: string
+    added: datetime
+    updated: datetime
+  }
+  ```
+
+- GET `/api/v1/package/:package/releases?page=:number`
+  ```
+  data: [<release>] // (x10)
+  ```
+- POST `/api/v1/package/:package/releases`
+  ```
+  data: <release>
+  ```
+- PATCH `/api/v1/package/:package/releases/:id`
+  ```
+  data: <release>
+  ```
+- DELETE `/apk/v1/package/:package/releases/:id`
+
+---
+
+### Tags
+
+Return a list of all tags for autocomplete for search.
+
+- GET `/api/v1/tags`
+  ```
+  items: [<tag>]
+  ```
+
+---
+
+### Authentication
+
+Is the user authenticated?
```

Comment:
I think here we can add more information if needed?

Reply:
Yeah I'm not sure what needs to go in here yet, But I will ping it, and if it 400's or something, they're not logged in, and if it returns a username, they are.

Probably some other github related things. avatar? userid? I don't actually know what github oauth will give you.

Reply:
Yes, if we need the other things GitHub returns, we can add it later.

Reply:
After another think, do we need to check the authentication in the backend? or we just need to it in the frontend?

When we hit the "login with github" on the [page](https://flink-ecosystem-demo.flink-resources.org/#/package/incredible-fresh-hat), we will not go to the backend.

Reply:
Github's oAuth setup does not allow for a front-end only solution.